scripts/deploy.js

```diff
@@ -0,0 +1,139 @@
+const packageJson = "../package.json";
+// Set to true to skip the download.
+const skipDownload = false;
+// Set to true to skip the upload.
+const skipUpload = false;
+
+const dataKey = "skynet-js";
+const versionsDir = "versions";
+const versionsTarFile = `${versionsDir}.tar`;
+
+(async () => {
+  const client = new SkynetClient("https://siasky.net");
+
+  // Get the latest version from package.json.
+
+  const version = require(packageJson).version;
+  console.log(`Version: ${version}`);
+
+  // Download the existing version directory.
+
+  if (fs.existsSync(versionsDir)) {
+    fs.rmSync(versionsDir, { recursive: true });
+  }
+  if (fs.existsSync(versionsTarFile)) {
+    fs.rmSync(versionsTarFile);
+  }
+
+  if (!skipDownload) {
+    try {
+      console.log(`Downloading HNS domain '${hnsDomain}' -> '${versionsTarFile}'`);
+      await client.downloadFileHns(versionsTarFile, hnsDomain, { format: "tar" });
+      // Untar to versions dir.
+      console.log(`Untarring '${versionsTarFile}' -> '${versionsDir}'`);
+      const writer = tar.extract(versionsDir, {
+        // Make sure all existing subfiles are readable.
+        readable: true,
+      });
+      await new Promise((resolve, reject) => {
+        fs.createReadStream(versionsTarFile).pipe(writer);
+        writer.on("finish", resolve);
+        writer.on("error", reject);
+      });
+
+      // Delete tar file.
+      fs.unlinkSync(versionsTarFile);
+    } catch (error) {
+      // If there was any error, stop. The initial directory should be uploaded manually.
+      console.log(error);
+      return;
+    }
+  }
+
+  // Copy the web bundle to the appropriate version dir.
+
+  // Compute the destination dir.
+  // TODO: Index by major version?
+  let versionSubdir = version.split(".").slice(0, 2).join(".");
+  const suffix = version.split("-").slice(1);
+  if (suffix.length > 0) {
+    versionSubdir = `${versionSubdir}-${suffix}`;
+  }
+  const destinationDir = `${versionsDir}/${versionSubdir}`;
+
+  // Copy the bundle. destination will be created or overwritten.
+  console.log(`Copying '${bundlePath}' -> '${destinationDir}'`);
+  if (fs.existsSync(destinationDir)) {
+    fs.rmSync(destinationDir, { recursive: true });
+  }
+  fs.mkdirSync(destinationDir, { recursive: true });
+  fse.copySync(bundlePath, destinationDir);
+
+  // Upload the directory and get the skylink.
+
+  if (!skipUpload) {
+    console.log(`Uploading '${versionsDir}' dir`);
+    let skylink = await client.uploadDirectory(versionsDir, { disableDefaultPath: true });
+    skylink = skylink.slice(uriSkynetPrefix.length);
+    console.log(`Skylink: ${skylink}`);
+
+    // Delete versionsDir.
+    fs.rmSync(versionsDir, { recursive: true });
+
+    // Update the registry entry.
+
+    console.log(`Updating '${dataKey}' registry entry with skylink`);
+    const seed = process.env[deploySeedEnvVar];
+    if (!seed) {
+      throw new Error(`Seed not found, make sure SKYNET_JS_DEPLOY_SEED is set`);
+    }
```

Comment:
nit: I'd put all critical validations at the start before we've done any I/O.

Reply:
Will do.